File: app/src/main/java/io/github/kowx712/mmuautoqr/utils/UserManager.java
package io.github.kowx712.mmuautoqr.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;
import io.github.kowx712.mmuautoqr.models.User;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class UserManager {
    private static final String PREFS_NAME = "AttendanceAppPrefs";
    private static final String KEY_USERS = "users";
    private static final String KEY_ENCRYPTION_KEY = "encryption_key";
    private static final String TAG = "UserManager";

    private final SharedPreferences prefs;
    private final Gson gson;

    public UserManager(Context context) {
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        gson = new Gson();

        // Get or generate encryption key
        String encryptionKey = prefs.getString(KEY_ENCRYPTION_KEY, null);
        if (encryptionKey == null) {
            encryptionKey = EncryptionUtils.generateKey();
            if (encryptionKey == null) {
                encryptionKey = "DefaultKey123";
            }
            prefs.edit().putString(KEY_ENCRYPTION_KEY, encryptionKey).apply();
        }
    }

    private void saveUsers(List<User> users) {
        try {
            List<User> usersToSave = new ArrayList<>();
            for (User user : users) {
                User userCopy = new User(user.getId(), user.getName(), user.getUserId(), user.getPassword());
                userCopy.setActive(user.isActive());

                String doubleEncrypted = EncryptionUtils.simpleEncrypt(
                        EncryptionUtils.simpleEncrypt(userCopy.getPassword())
                );
                userCopy.setPassword(doubleEncrypted);
                usersToSave.add(userCopy);
            }

            String json = gson.toJson(usersToSave);
            SharedPreferences.Editor editor = prefs.edit();
            editor.putString(KEY_USERS, json);
            editor.apply();
            editor = null;
        } catch (Exception e) {
            Log.e(TAG, "Error saving users", e);
        }
    }

    public List<User> getUsers() {
        String json = prefs.getString(KEY_USERS, null);
        if (json != null) {
            try {
                Type type = new TypeToken<List<User>>(){}.getType();
                List<User> users = gson.fromJson(json, type);

                if (users == null) {
                    return new ArrayList<>();
                }

                for (User user : users) {
                    if (user.getPassword() != null && !user.getPassword().isEmpty()) {
                        try {
                            String firstDecrypt = EncryptionUtils.simpleDecrypt(user.getPassword());
                            String finalDecrypt = EncryptionUtils.simpleDecrypt(firstDecrypt);
                            user.setPassword(finalDecrypt);
                        } catch (Exception e) {
                            try {
                                String singleDecrypt = EncryptionUtils.simpleDecrypt(user.getPassword());
                                user.setPassword(singleDecrypt);
                            } catch (Exception ex) {
                                Log.e(TAG, "Error decrypting password (inner catch)", ex);
                            }
                        }
                    }
                }
                return users;
            } catch (Exception e) {
                Log.e(TAG, "Error getting users", e);
                return new ArrayList<>();
            }
        }
        return new ArrayList<>();
    }

    public boolean addUser(String name, String userId, String password) {
        List<User> users = getUsers();

        for (User user : users) {
            if (user.getUserId().equals(userId)) {
                return false;
            }
        }

        User newUser = new User(UUID.randomUUID().toString(), name, userId, password);
        users.add(newUser);
        saveUsers(users);
        return true;
    }

    public boolean updateUser(String userId, String name, String password) {
        List<User> users = getUsers();

        for (User user : users) {
            if (user.getUserId().equals(userId)) {
                user.setName(name);
                user.setPassword(password);
                saveUsers(users);
                return true;
            }
        }
        return false;
    }

    public boolean deleteUser(String userId) {
        List<User> users = getUsers();

        for (int i = 0; i < users.size(); i++) {
            if (users.get(i).getUserId().equals(userId)) {
                users.remove(i);
                saveUsers(users);
                return true;
            }
        }
        return false;
    }

    public List<User> getActiveUsers() {
        List<User> allUsers = getUsers();
        List<User> activeUsers = new ArrayList<>();

        for (User user : allUsers) {
            if (user.isActive()) {
                activeUsers.add(user);
            }
        }
        return activeUsers;
    }

    public void toggleUserStatus(String userId) {
        List<User> users = getUsers();

        for (User user : users) {
            if (user.getUserId().equals(userId)) {
                user.setActive(!user.isActive());
                saveUsers(users);
                break;
            }
        }
    }

    public void clearAllUsers() {
        prefs.edit().remove(KEY_USERS).apply();
    }

    public int getUserCount() {
        return getUsers().size();
    }

    public int getActiveUserCount() {
        return getActiveUsers().size();
    }
}

File: app/src/main/java/io/github/kowx712/mmuautoqr/utils/UserManager.kt
package io.github.kowx712.mmuautoqr.utils

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.core.content.edit
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import io.github.kowx712.mmuautoqr.models.User
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class UserManager(context: Context) {
    private val sharedPreferences: SharedPreferences =
        context.getSharedPreferences("user_prefs", Context.MODE_PRIVATE)
    private val gson = Gson()

    private var _cachedUsers: MutableList<User>? = null

    private companion object {
        private const val TAG = "UserManager"
        private const val ENCRYPTION_KEY_PREFS_KEY = "user_manager_encryption_key_v1"
        private const val USERS_LIST_PREFS_KEY = "users_list_encrypted_v1"
    }

    private val encryptionKey: String

    init {
        val storedKey = sharedPreferences.getString(ENCRYPTION_KEY_PREFS_KEY, null)
        if (storedKey != null) {
            this.encryptionKey = storedKey
        } else {
            val generatedKey = EncryptionUtils.generateKey()
            if (generatedKey != null) {
                this.encryptionKey = generatedKey
                sharedPreferences.edit { putString(ENCRYPTION_KEY_PREFS_KEY, generatedKey) }
            } else {
                Log.e(TAG, "Failed to generate encryption key for UserManager. Data security cannot be ensured.")
                throw IllegalStateException("Failed to generate encryption key for UserManager. Data security cannot be ensured.")
            }
        }
    }

    suspend fun getUsers(): List<User> = withContext(Dispatchers.IO) {
        _cachedUsers?.toList()?.let { return@withContext it }

        val encryptedJson = sharedPreferences.getString(USERS_LIST_PREFS_KEY, null)
        val loadedUsers: MutableList<User> = if (encryptedJson != null && encryptedJson.isNotEmpty()) {
            val json = EncryptionUtils.decrypt(encryptedJson, encryptionKey)
            val type = object : TypeToken<MutableList<User>>() {}.type
            try {
                gson.fromJson(json, type) ?: mutableListOf()
            } catch (e: Exception) {
                Log.e(TAG, "Error parsing users JSON from SharedPreferences. Clearing user data.", e)
                clearAllUsers()
                mutableListOf()
            }
        } else {
            mutableListOf()
        }
        _cachedUsers = loadedUsers
        loadedUsers.toList()
    }

    suspend fun getUserCount(): Int = getUsers().size

    suspend fun getActiveUserCount(): Int = getUsers().count { it.isActive }

    private suspend fun saveUsersToPrefs(usersToSave: List<User>) = withContext(Dispatchers.IO) {
        val json = gson.toJson(usersToSave)
        val encryptedJson = EncryptionUtils.encrypt(json, encryptionKey)
        sharedPreferences.edit { putString(USERS_LIST_PREFS_KEY, encryptedJson) }
    }

    suspend fun addUser(name: String, userId: String, password: String): Boolean {
        val currentUsers = _cachedUsers ?: getUsers().toMutableList()
        if (currentUsers.any { it.userId == userId }) {
            return false
        }
        val newUser = User(name = name, userId = userId, password = password)
        currentUsers.add(newUser)
        _cachedUsers = currentUsers
        saveUsersToPrefs(currentUsers)
        return true
    }

    suspend fun updateUser(userIdToUpdate: String, newName: String, newPassword: String): Boolean {
        val currentUsers = _cachedUsers ?: getUsers().toMutableList()
        val userIndex = currentUsers.indexOfFirst { it.userId == userIdToUpdate }
        if (userIndex != -1) {
            val oldUser = currentUsers[userIndex]
            currentUsers[userIndex] = oldUser.copy(name = newName, password = newPassword)
            _cachedUsers = currentUsers
            saveUsersToPrefs(currentUsers)
            return true
        }
        return false
    }

    suspend fun deleteUser(userIdToDelete: String): Boolean {
        val currentUsers = _cachedUsers ?: getUsers().toMutableList()
        val removed = currentUsers.removeAll { it.userId == userIdToDelete }
        if (removed) {
            _cachedUsers = currentUsers
            saveUsersToPrefs(currentUsers)
        }
        return removed
    }

    suspend fun toggleUserStatus(userIdToToggle: String): Boolean {
        val currentUsers = _cachedUsers ?: getUsers().toMutableList()
        val userIndex = currentUsers.indexOfFirst { it.userId == userIdToToggle }
        if (userIndex != -1) {
            val oldUser = currentUsers[userIndex]
            currentUsers[userIndex] = oldUser.copy(isActive = !oldUser.isActive)
            _cachedUsers = currentUsers
            saveUsersToPrefs(currentUsers)
            return true
        }
        return false
    }

    suspend fun clearAllUsers() {
        _cachedUsers = mutableListOf()
        saveUsersToPrefs(emptyList())
    }

    fun refreshCache() {
        _cachedUsers = null
    }
}
